sql/log.cc

```diff
@@ -12914,3 +13024,221 @@ void wsrep_register_binlog_handler(THD *thd, bool trx)
+    DBUG_ASSERT(!m_replaced);
+    DBUG_ASSERT(!mysql_bin_log.is_open());
+  }
+
+  DBUG_EXECUTE_IF("binlog_cache_free_flush_crash_after_rotate",
+                  DBUG_SUICIDE(););
+
+  if (!m_replaced)
+  {
+    mysql_mutex_unlock(&mysql_bin_log.LOCK_log);
+    if (check_purge)
+      mysql_bin_log.checkpoint_and_purge(prev_binlog_id);
+    return false;
+  }
+
+  mysql_bin_log.trx_group_commit_leader(entry, true /* is_free_flush */);
+
+  if (check_purge)
+    mysql_bin_log.checkpoint_and_purge(prev_binlog_id);
+  return true;
+}
+
+bool Binlog_cache_free_flush::replace_binlog_file()
+{
+  size_t binlog_size= my_b_tell(&mysql_bin_log.log_file);
+  size_t required_size= binlog_size;
+  // space for Gtid_log_event
+  required_size+= LOG_EVENT_HEADER_LEN + Gtid_log_event::max_data_length +
+                  BINLOG_CHECKSUM_LEN;
+
+  if (DBUG_IF("simulate_required_size_too_big") ||
+      required_size > m_cache_data->file_reserved_bytes())
+  {
+    sql_print_information("Could not rename binlog cache to binlog, "
+                          "require %llu bytes but only %llu bytes reserved.",
+                          required_size, m_cache_data->file_reserved_bytes());
+    return false;
+  }
+  char temp_name[FN_REFLEN];
+  File read_fd= -1;
+  File new_log_fd= -1;
+  bool ret= false;
+  size_t copy_len= 0;
+  uchar buf[IO_SIZE];
+
+  m_cache_data->tmp_file_name(temp_name);
+
+  read_fd= mysql_file_open(key_file_binlog, mysql_bin_log.get_log_fname(),
+                           O_RDONLY | O_BINARY | O_SHARE, MYF(MY_WME));
+  /* Create fd for the cache file as a new binlog file fd */
+  new_log_fd= mysql_file_open(key_file_binlog, temp_name,
+                              O_BINARY | O_CLOEXEC | O_WRONLY, MYF(MY_WME));
+  if (read_fd == -1 || new_log_fd == -1)
+    goto end;
+
+  /* copy the content of binlog file into the binlog cache's header */
+  while (copy_len < binlog_size)
+  {
+    int read_len= (int) mysql_file_read(read_fd, buf, IO_SIZE, MYF(MY_WME));
+    if (read_len < 0 ||
+        mysql_file_write(new_log_fd, buf, read_len,
+                         MYF(MY_WME | MY_NABP | MY_WAIT_IF_FULL)))
+      goto end;
+
+    copy_len+= read_len;
+  }
+
+  if (DBUG_IF("simulate_write_binlog_cache_head_error") ||
+      mysql_file_sync(new_log_fd, MYF(MY_WME)))
+    goto end;
+
+  my_delete(mysql_bin_log.get_log_fname(), MYF(0));
+
+  if (DBUG_IF("simulate_rename_binlog_cache_to_binlog_error") ||
+      my_rename(temp_name, mysql_bin_log.get_log_fname(), MYF(MY_WME)))
+  {
+    /*
+      Since the binlog file is deleted, it should return true if error
+      happens. Thus the index file will be reverted later.
+    */
+    ret= true;
+    goto end;
+  }
+
+  sql_print_information("Renamed binlog cache to binlog %s",
+                        mysql_bin_log.get_log_fname());
+
+  mysql_file_close(mysql_bin_log.log_file.file, MYF(MY_WME));
+  mysql_bin_log.log_file.file= new_log_fd;
+  new_log_fd= -1;
+
+  if (mysql_bin_log.write_gtid_event(
```

Comment:
I still think this should happen before the actual replacement. I imagine you can create an `IO_CACHE` of the temp file, and pass it in as a new optional parameter to `write_gtid_event`, which in-turn would use 
```
bool Event_log::write_event(Log_event *ev, binlog_cache_data *data,
                            IO_CACHE *file)
```

if the parameter is not null. Then I imagine the copying of the header events could use the IO cache too, and then everything would be flushed together.

Reply:
moved it before actual replacement.